firedrake/slate/slate.py

```diff
@@ -0,0 +1,749 @@
+    id = 0
+
+    def __init__(self, arguments, coefficients, integrals):
+        """Constructor for the Tensor class."""
+        self.id = Tensor.id
+        Tensor.id += 1
+        self._arguments = arguments
+        self._coefficients = coefficients
+        self._integrals = _sorted_integrals(integrals)
+
+        slate_integrals = []
+        for integral in self._integrals:
+            slate_integrals.append(SlateIntegral(integral))
+        self._slate_integrals = tuple(slate_integrals)
+
+        # Compute Tensor shape information
+        shape = []
+        shapes = {}
+
+        for i, arg in enumerate(self._arguments):
+            V = arg.function_space()
+            shapelist = []
+
+            for fs in V:
+                shapelist.append(fs.fiat_element.space_dimension() * fs.dim)
+
+            shapes[i] = tuple(shapelist)
+            shape.append(sum(shapelist))
+
+        self.shapes = shapes
+        self.shape = tuple(shape)
+        self.rank = len(self._arguments)
+
+        # Compute integration domains
+        integral_domains = join_domains([it.ufl_domain() for it in self._integrals])
+        self._integral_domains = sort_domains(integral_domains)
+
+        # Compute relevant attributes for signature computation
+        self._coefficient_numbering = dict((c, i) for i, c in enumerate(self._coefficients))
+        self._domain_numbering = dict((d, i) for i, d in enumerate(self._integral_domains))
+
+        self._hash = None
+        self._signature = None
+
+    # Overloaded operations for the tensor algebra
+    def __add__(self, other):
+        return TensorAdd(self, other)
+
+    def __radd__(self, other):
+        """Ordering of tensor addition does not matter."""
+        return self.__add__(other)
+
+    def __sub__(self, other):
+        return TensorSub(self, other)
+
+    def __rsub__(self, other):
+        """Ordering of tensor subtraction does not matter."""
+        return self.__sub__(other)
+
+    def __mul__(self, other):
+        return TensorMul(self, other)
+
+    def __rmul__(self, other):
+        """Tensor multiplication is not commutative in general."""
+        return TensorMul(other, self)
+
+    def __neg__(self):
+        return Negative(self)
+
+    def __pos__(self):
+        return Positive(self)
+
+    def equals(self, other):
+        """Evaluates the boolean expression `A == B`."""
+
+        if not isinstance(other, Tensor):
+            return False
+        if len(self.arguments()) != len(other.arguments()):
+            return False
+        if len(self.coefficients()) != len(other.coefficients()):
+            return False
+        if len(self._integrals) != len(other._integrals):
+            return False
+        return (self._arguments == other._arguments and
+                self._coefficients == other._coefficients and
+                self._integrals == other._integrals)
+
+    def __ne__(self, other):
+        return not self.equals(other)
```

Comment:
This is a pattern copied from UFL, that allows one to write "a == L" as an equation... Which is sort of ugh.